src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -64126,6 +64129,339 @@ void local2()
+            CreateCompilation(source).VerifyDiagnostics(
+                // (15,13): warning CS8602: Dereference of a possibly null reference.
+                //             F.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F").WithLocation(15, 13),
+                // (16,13): warning CS8602: Dereference of a possibly null reference.
+                //             s.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(16, 13));
+        }
+
+        [Theory, CombinatorialData, WorkItem("https://github.com/dotnet/roslyn/issues/71419")]
+        public void LocalFunction_UnreachableGraph_01(bool allocateSlot)
+        {
+            var source = $$"""
+                #nullable enable
+                class C
+                {
+                    string? F;
+                    void M()
+                    {
+                        string? s = null;
+                        {{(allocateSlot ? "F = null;" : "")}}
+
+                        void local1()
+                        {
+                            for (int i = 0; i < 3; i++) { }
+
+                            F.ToString();
+                            s.ToString();
+                        }
+
+                        void local2()
+                        {
+                            F = "a";
+                            s = "b";
+                            local1();
+                        }
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (18,14): warning CS8321: The local function 'local2' is declared but never used
+                //         void local2()
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "local2").WithArguments("local2").WithLocation(18, 14));
+        }
+
+        [Theory, CombinatorialData, WorkItem("https://github.com/dotnet/roslyn/issues/71419")]
+        public void LocalFunction_UnreachableGraph_02(bool allocateSlot)
+        {
+            var source = $$"""
+                #nullable enable
+                class C
+                {
+                    string? F;
+                    void M()
+                    {
+                        string? s = null;
+                        {{(allocateSlot ? "F = null;" : "")}}
+
+                        void local1()
+                        {
+                            F = "a";
+                            s = "b";
+                            local2();
+                        }
+
+                        void local2()
+                        {
+                            for (int i = 0; i < 3; i++) { }
+
+                            F.ToString();
+                            s.ToString();
+                        }
+                    }
+                }
+                """;
+            // Notice the warning which is not present when the local funcs are swapped (cf. LocalFunction_UnreachableGraph_01).
```

Comment:
>// Notice the warning which is not present when the local funcs are swapped (cf. LocalFunction_UnreachableGraph_01). [](http://example.com/codeflow?start=12&length=116)

I would actually expect an opposite behavior between between the tests. Specifically, I would expect us to visit "never used" local functions in lexical order and report possible null dereference in ```LocalFunction_UnreachableGraph_01``` instead. It looks like that would also agree with the "never used" warning. Since we didn't report it for 'local2', we must have considered it "used" through 'local1' and that one initializes `F` to not null value before calling 'local2'.

Reply:
This swapped, but it still doesn't agree with the "never used" warning in equivalent way. In the other test we also don't report the "never used" warning for one function used through the other which initializes `F` to non-null value, but we still report the nullable warning.

Reply:
> but it still doesn't agree with the "never used" warning in equivalent way. In the other test we also don't report the "never used" warning for one function used through the other which initializes F to non-null value, but we still report the nullable warning.

The consistency with the  "never used" warning was not my primary motivation. The primary motivation is that I think that "unreachable" local functions should be visited in lexical order (not in reversed order), the order is observable and lexical order will be more intuitive for a user.